packages/action-listener-middleware/src/index.ts

```diff
@@ -65,73 +67,109 @@ function assertFunction(
 const defaultWhen: MiddlewarePhase = 'afterReducer'
 const actualMiddlewarePhases = ['beforeReducer', 'afterReducer'] as const
 
+function assertActive(signal: AbortSignal, reason?: string) {
+  if (signal.aborted) {
+    throw new TaskAbortError(reason)
+  }
+}
+
+function createDelay(signal: AbortSignal) {
+  return function delay(timeoutMs: number): Promise<void> {
+    return new Promise((resolve, reject) => {
+      assertActive(signal)
+
+      setTimeout(() => {
+        try {
+          assertActive(signal)
+        } catch (err) {
+          reject(err)
+        }
+        resolve()
+      }, timeoutMs)
+    })
+  }
+}
```

Comment:
```suggestion
async function assertActive(signal: AbortSignal, reason?: string) {
  if (signal.aborted) {
    throw new TaskAbortError(reason)
  }
}

function createDelay(signal: AbortSignal) {
  return async function delay(timeoutMs: number): Promise<void> {
    await assertActive(signal)
    await new Promise((resolve) =>resolve, timeoutMs)
    await assertActive(signal)
  }
}
```

Might be worth exploring to write `assertActive` as an `async` function - at least `createDelay` would become significantly shorter. Not sure about the more complex functions though.

Reply:
Otoh... it wouldn't even need to be async for that change in `createDelay` 🤔

Reply:
I'm microoptimizing again.. 😅  this is looking very good. I'm not really awake enough to perfectly understand everything yet (I'm not on top with the full PR). but what I understand goes into a great direction 👍

Reply:
> I'm microoptimizing again.. 😅 this is looking very good. I'm not really awake enough to perfectly understand everything yet (I'm not on top with the full PR). but what I understand goes into a great direction 👍

Done 👍